packages/components/src/autocomplete/types.ts

```diff
@@ -0,0 +1,213 @@
+/**
+ * WordPress dependencies
+ */
+import type { WPElement } from '@wordpress/element';
+/**
+ * Internal dependencies
+ */
+import type { useAutocomplete } from '.';
+
+type GetOptions< TCompleterOption = any > = (
+	filterValue: AutocompleterUIProps[ 'filterValue' ]
+) => Array< TCompleterOption > | Promise< Array< TCompleterOption > >;
+
+export type OptionCompletion = {
+	action: 'insert-at-caret' | 'replace';
+	value: string | WPElement;
+};
```

Comment:
On definitely typed, the types for `OptionCompletion` differ a little

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/wordpress__components/autocomplete/index.d.ts#L13-L16

I think they are actually a cleaner solution, as they also allow to type the return type of `getOptionCompletion` simply as `OptionCompletion` https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/wordpress__components/autocomplete/index.d.ts#L79

Reply:
Good point, overall it looks like a good solution to me.

While checking with the types from definitely typed, I noticed another difference: when the `action ` is 'replace', the types from definitely typed say [that `value` should be of type `Value`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/wordpress__components/autocomplete/index.d.ts#L16), which in this PR you called `RichTextValue`.

But when I tried type `OptionCompletion` as

```ts
export type OptionCompletion =
	| { action: 'insert-at-caret'; value: React.ReactNode }
	| { action: 'replace'; value: RichTextValue };
```

TypeScript complains in the `insertCompletion` function, saying that a `RichTextValue` object can't be passed to the `renderToString` funciton.

https://github.com/WordPress/gutenberg/blob/8c17bb9e488b7ba76ed764fbef257257c9846552/packages/components/src/autocomplete/index.tsx

We need to understand where is the issue:
- either the definitely typed types are out of date, and therefore Chad's types for the `value` property are correct
- or, if the definitely typed types are correct, we'd need to improve the code in the `insertCompletion` function.

Reply:
Tbh, I'm not sure how the completers API works, I think the best person to ask is @draganescu. The `React.ReactNode` type doesn't make any sense to me, it's should probably either be a rich text value or just a string?

Reply:
Sorry, not @draganescu. I meant @brandonpayton.

Reply:
Alright, I tinkered a bit with the types and I think I should have landed on a good solution:

- Refactored how the `OptionCompletion` type is declaired, extracting separate `InsertOption` and `ReplaceOption` types, and adding `React.ReactNode` as well
- Refined the `onReplace` callback to only accept `RichTextValue[]` as argument, since it seems to be called only with `ReplaceOption` objects
- Refined the `insertCompletion` function to only accept `ReactNode` as argument, since it seems to be only called with `InsertOption` objects
- Rewrote the `isCompletionObject` typeguard to also check against `ReactNode`, and to more explicitly return only `InsertOption | ReplaceOption`
- Removed the `{action, value}` object destructuring — referring the whole `completionObject` variable actually helps TypeScript to narrow types in the `if` statements

<details>

<summary>Here's the diff</summary>

```diff
diff --git a/packages/components/src/autocomplete/index.tsx b/packages/components/src/autocomplete/index.tsx
index a7f86d1a1c..e62d018e7d 100644
--- a/packages/components/src/autocomplete/index.tsx
+++ b/packages/components/src/autocomplete/index.tsx
@@ -43,6 +43,8 @@ import type {
 	OptionCompletion,
 	UseAutocompleteProps,
 	WPCompleter,
+	InsertOption,
+	ReplaceOption,
 } from './types';
 
 export function useAutocomplete( {
@@ -68,7 +70,7 @@ export function useAutocomplete( {
 	>( null );
 	const backspacing = useRef( false );
 
-	function insertCompletion( replacement: OptionCompletion[ 'value' ] ) {
+	function insertCompletion( replacement: React.ReactNode ) {
 		if ( autocompleter === null ) {
 			return;
 		}
@@ -90,26 +92,31 @@ export function useAutocomplete( {
 		if ( getOptionCompletion ) {
 			const completion = getOptionCompletion( option.value, filterValue );
 
-			function isCompletionObject(
-				completionData: any
-			): completionData is OptionCompletion {
+			const isCompletionObject = (
+				obj: OptionCompletion
+			): obj is InsertOption | ReplaceOption => {
 				return (
-					undefined !== completionData.action &&
-					undefined !== completionData.value
+					obj !== null &&
+					typeof obj === 'object' &&
+					'action' in obj &&
+					'value' in obj
 				);
-			}
+			};
 
-			const { action, value } = isCompletionObject( completion )
+			const completionObject = isCompletionObject( completion )
 				? completion
-				: { action: 'insert-at-caret', value: completion };
+				: ( {
+						action: 'insert-at-caret',
+						value: completion,
+				  } as InsertOption );
 
-			if ( 'replace' === action ) {
-				onReplace( [ value ] );
+			if ( 'replace' === completionObject.action ) {
+				onReplace( [ completionObject.value ] );
 				// When replacing, the component will unmount, so don't reset
 				// state (below) on an unmounted component.
 				return;
-			} else if ( 'insert-at-caret' === action ) {
-				insertCompletion( value );
+			} else if ( 'insert-at-caret' === completionObject.action ) {
+				insertCompletion( completionObject.value );
 			}
 		}
 
diff --git a/packages/components/src/autocomplete/types.ts b/packages/components/src/autocomplete/types.ts
index 6783808683..f318322ed6 100644
--- a/packages/components/src/autocomplete/types.ts
+++ b/packages/components/src/autocomplete/types.ts
@@ -7,11 +7,17 @@ import type { WPElement } from '@wordpress/element';
  */
 import type { useAutocomplete } from '.';
 
-export type OptionCompletion = {
-	action: 'insert-at-caret' | 'replace';
-	value: string | WPElement;
+// Insert the `value` into the text.
+export type InsertOption = {
+	action: 'insert-at-caret';
+	value: React.ReactNode;
 };
 
+// Replace the current block with the block specified in the `value` property
+export type ReplaceOption = { action: 'replace'; value: RichTextValue };
+
+export type OptionCompletion = React.ReactNode | InsertOption | ReplaceOption;
+
 type OptionLabel = string | WPElement | Array< string | WPElement >;
 export type KeyedOption = {
 	key: string;
@@ -77,7 +83,7 @@ export type WPCompleter< TCompleterOption = any > = {
 	getOptionCompletion?: (
 		option: TCompleterOption,
 		query: string
-	) => OptionCompletion | OptionCompletion[ 'value' ];
+	) => OptionCompletion;
 	/**
 	 * A function that returns an array of items to be displayed in the
 	 * Autocomplete UI. These items have uniform shape and have been filtered by
@@ -185,7 +191,7 @@ export type UseAutocompleteProps = {
 	 * A function to be called when an option is selected to replace the
 	 * existing text.
 	 */
-	onReplace: ( arg: [ OptionCompletion[ 'value' ] ] ) => void;
+	onReplace: ( values: RichTextValue[] ) => void;
 	/**
 	 * An array of all of the completers to apply to the current element.
 	 */
```
</details>

What do y'all think ?

Reply:
Why allow `React.ReactNode`? This shouldn't be possible. I think probably only strings and rich text values are allowed, but surely no React nodes?

Reply:
I was assuming `ReactNode` because [that value is passed to the `renderToString` function](https://github.com/WordPress/gutenberg/blob/8c17bb9e488b7ba76ed764fbef257257c9846552/packages/components/src/autocomplete/index.tsx#L78), which accepts a `ReactNode`

<details>

<summary>But I'm totally fine with replacing `ReactNode` with `string`</summary>

```diff
diff --git a/packages/components/src/autocomplete/index.tsx b/packages/components/src/autocomplete/index.tsx
index a7f86d1a1c..4275fed1c3 100644
--- a/packages/components/src/autocomplete/index.tsx
+++ b/packages/components/src/autocomplete/index.tsx
@@ -43,6 +43,8 @@ import type {
 	OptionCompletion,
 	UseAutocompleteProps,
 	WPCompleter,
+	InsertOption,
+	ReplaceOption,
 } from './types';
 
 export function useAutocomplete( {
@@ -68,7 +70,7 @@ export function useAutocomplete( {
 	>( null );
 	const backspacing = useRef( false );
 
-	function insertCompletion( replacement: OptionCompletion[ 'value' ] ) {
+	function insertCompletion( replacement: string ) {
 		if ( autocompleter === null ) {
 			return;
 		}
@@ -90,26 +92,33 @@ export function useAutocomplete( {
 		if ( getOptionCompletion ) {
 			const completion = getOptionCompletion( option.value, filterValue );
 
-			function isCompletionObject(
-				completionData: any
-			): completionData is OptionCompletion {
+			const isCompletionObject = (
+				obj: OptionCompletion
+			): obj is InsertOption | ReplaceOption => {
 				return (
-					undefined !== completionData.action &&
-					undefined !== completionData.value
+					// Since `typeof null === 'object'`,
+					// we need to check explicitely for `null`
+					obj !== null &&
+					typeof obj === 'object' &&
+					'action' in obj &&
+					'value' in obj
 				);
-			}
+			};
 
-			const { action, value } = isCompletionObject( completion )
+			const completionObject = isCompletionObject( completion )
 				? completion
-				: { action: 'insert-at-caret', value: completion };
+				: ( {
+						action: 'insert-at-caret',
+						value: completion,
+				  } as InsertOption );
 
-			if ( 'replace' === action ) {
-				onReplace( [ value ] );
+			if ( 'replace' === completionObject.action ) {
+				onReplace( [ completionObject.value ] );
 				// When replacing, the component will unmount, so don't reset
 				// state (below) on an unmounted component.
 				return;
-			} else if ( 'insert-at-caret' === action ) {
-				insertCompletion( value );
+			} else if ( 'insert-at-caret' === completionObject.action ) {
+				insertCompletion( completionObject.value );
 			}
 		}
 
diff --git a/packages/components/src/autocomplete/types.ts b/packages/components/src/autocomplete/types.ts
index 6783808683..bfdd826031 100644
--- a/packages/components/src/autocomplete/types.ts
+++ b/packages/components/src/autocomplete/types.ts
@@ -7,11 +7,17 @@ import type { WPElement } from '@wordpress/element';
  */
 import type { useAutocomplete } from '.';
 
-export type OptionCompletion = {
-	action: 'insert-at-caret' | 'replace';
-	value: string | WPElement;
+// Insert the `value` into the text.
+export type InsertOption = {
+	action: 'insert-at-caret';
+	value: string;
 };
 
+// Replace the current block with the block specified in the `value` property
+export type ReplaceOption = { action: 'replace'; value: RichTextValue };
+
+export type OptionCompletion = string | InsertOption | ReplaceOption;
+
 type OptionLabel = string | WPElement | Array< string | WPElement >;
 export type KeyedOption = {
 	key: string;
@@ -77,7 +83,7 @@ export type WPCompleter< TCompleterOption = any > = {
 	getOptionCompletion?: (
 		option: TCompleterOption,
 		query: string
-	) => OptionCompletion | OptionCompletion[ 'value' ];
+	) => OptionCompletion;
 	/**
 	 * A function that returns an array of items to be displayed in the
 	 * Autocomplete UI. These items have uniform shape and have been filtered by
@@ -185,7 +191,7 @@ export type UseAutocompleteProps = {
 	 * A function to be called when an option is selected to replace the
 	 * existing text.
 	 */
-	onReplace: ( arg: [ OptionCompletion[ 'value' ] ] ) => void;
+	onReplace: ( values: RichTextValue[] ) => void;
 	/**
 	 * An array of all of the completers to apply to the current element.
 	 */

```

</details>

Reply:
It looks like that was introduced in #2896. Mentions are now just inserting a string though, not a link. Since we're only using it with strings and we don't have it documented anywhere, I would prefer removing it.

Reply:
Ugh, looks like it is documented. Ok, never mind :)

Reply:
Just to make sure I understand your train of thoughts — your preference was to replace `ReactNode` with `string`, but since it's documented we should keep `ReactNode` in place ?

Reply:
Wow, thanks @ciampo. It looks solid to me! Just to make sure I'm on the right page before applying the recommended change, we're going to stick with allowing `React.ReactNode`, correct?

Reply:
Yeah I think so. We can always open a quick follow-up if it turns our that `string` was a better option

Reply:
Yes, ReactNode is fine. Wasn't aware that we're using it.

Reply:
Thank you both - I've applied @ciampo's recommended changes, and rebased the PR to include the changes to the component from other recent updates.